opentelemetry/proto/collector/logs/v1/logs_service.proto

```diff
@@ -43,4 +43,18 @@ message ExportLogsServiceRequest {
 }
 
 message ExportLogsServiceResponse {
+  // The details of a partially successful export request.
+  // When the request is only partially successful (due to e.g. invalid/bad data),
+  // the server MUST populate the partial_success property with a
+  // developer-facing error message along with the number of accepted log records.
+  ExportLogsPartialSuccess partial_success = 1;
```

Comment:
If that is not the case then let's make it optional

Reply:
I agree. @joaopgrassi made the point in in https://github.com/open-telemetry/opentelemetry-proto/pull/390#discussion_r885831423, so we should call these optional since presence/absence is considered meaningful.

Reply:
Hi all, I'm not entirely sure to which field you are referring to? The `partial_success` one or the numeric field with the number of accepted records? 

If it's for the non-primitive `partial_success` field, I'm not sure if there's any gain in marking it optional? Since it's not a primitive type, we will not have the _explicit presence_ semantics on it, right? I'm not a proto expert but that's how I understand it. 

If you are talking about the numeric field, I also think we should not mark it as optional. The idea of the new `partial_success` field is to **only** be populated in cases where there was a partial success AND the server is able to report it (processing of the request is not done at a later point for ex). 

In that case, the idea I believe is to make it a MUST in the spec. Meaning, `partial_success` will always be initialized and the numeric field containing the accepted records is always there. So `0` there would really mean 0 records were accepted. 

For the "total success case" e.g. when ALL records were accepted by the server, OR the server does the processing later thus can't know how many records were accepted, the `partial_success` should NOT be set, so it will be null.

Does this make sense? I put these requirements in the spec change here https://github.com/open-telemetry/opentelemetry-specification/pull/2636

Reply:
For example, I can change the comments here to reflect the spec text more or less:

```
  // The details of a partially successful export request.
  // If the processing of the request is partially successful
  // (i.e. when the server accepts only parts of the data and rejects the rest)
  // and the number of accepted log records is known, the server MUST initialize
  // the partial_success field and MUST set the accepted_log_records.

  // When the export request is processed successfully, the server MUST leave
  // the partial_success field unset.
```

Reply:
@bogdandrutu as we talked, I added some doc on `error_message` stating that an empty string is equivalent of it not being set at all, to avoid the overhead. I don't think it's important to distinguish between it "ever being set" from "empty string". Thanks for the pointer. [f299099](https://github.com/open-telemetry/opentelemetry-proto/pull/390/commits/f29909964eaf2885c2a8a8c831c97dc80aa5a861)

Reply:
@jmacd is that also what you meant? Or did I miss something else?

Reply:
I believe the question here is whether when `error_message` is set to a non-empty string AND `accepted_things` equals 0, what does it mean? Since the error message is set and number of accepted things is not optional, the implication is that 0 is meaningful, i.e., no data was accepted. Thus if you are going to set an error message, you MUST set the number of accepted things. Is this true?

Reply:
Or, @joaopgrassi is the implication that if no "things" were accepted, there is no "partial" success, and an RPC error should be returned instead? That seems awkward to me: if every point drops, I'd like to return a partial success with 0 accepted things.

Reply:
@jmacd 

>if every point drops, I'd like to return a partial success with 0 accepted things.

Yes, that is the approach we are going for here.

The idea is: If there's a partial success (meaning, the server knows before responding to clients how many "things" it accepted and how many it rejected) then the server MUST initialize the `partial_success` message and it MUST set the number in the `accepted_xxx` field. Here then `0` is meaningful. The server dropped all spans/points/log records. I think it's important that this case is supported like this instead of relying on RPC error.

The server MAY also set the `error_message` but that's not required. If `error_message` is `""`, it means it was either empty by intention or never set. There's no distinction in this case. Since we don't offer guidance on what to put in the message, making it a MAY makes sense to me.

I wrote all of these nuances in the spec. I would appreciate if you all could take a look at it. Hopefully it clarifies all the questions brought up here: https://github.com/open-telemetry/opentelemetry-specification/pull/2636

Reply:
Three possibilities:
1. Make each `*PartialSuccess` field optional
2. Make each `accepted_*` field optional
3. Require error_details to be non-empty when `accepted_*` is meaningfully zero

I prefer number 3.

Reply:
@jmacd I believe the original intention was: 

If there was a "complete success" OR the server does the processing later, thus can't know how many things it will accept, then the `PartialSucess` **must** be left unset.

The `PartialSuccess` **must only be set** when there's in fact a partial success on the server. I think this way it's easier to reason about than making x field required IF y field is present with some value. `0` then means nothing was accepted and there's no other alternative interpretation. 

All of this is explained in the spec PR I have open: https://github.com/open-telemetry/opentelemetry-specification/pull/2636

Reply:
Google has written an explainer on field presence for proto2 and proto3, https://github.com/protocolbuffers/protobuf/blob/main/docs/field_presence.md,

The implication (under proto3) is that you cannot set an `error_details` string without causing the whole message to be serialized, at which point it will look like the `accepted_things` field is present. So behavior (3) described by me is equivalent to number (1) as long as google doesn't go changing the definition of field presence again.

By these rules, you cannot encode a meaningful 0-valued `accepted_things` inside a message with an empty `error_details`, because it is equivalent to the serialized form of a not-present field; the deserializer seeing a `0` and an empty string will leave you a null and your meaningful-0 disappears.

To ensure that we get a `null` when the field is not set, we need to use the optional keyword. I think we should go with option 1 (functionally equivalent to 3).

Reply:
@bogdandrutu do you agree?

Reply:
@jmacd 

Let's make sure we are in the same page for the "total success" case. In this, the new message here will not be set, as per spec so we are good. This is how we will distinguish a "total success" case from a "partial success". Also, it's not a primitive type, so there's no point in making that `optional`.

Now, if I got your point across, your concern is: There's a partial success. The server _can_ instantiate the `partial_success` sub message and leave everything as is with their default values. With this, it will not be possible to tell if `0` was actually set or if it was just the default?

If this is the case, then I think making `int64 accepted_*` as `optional` would solve it, no? The spec will say that when `partial_success` is initialized, the server MUST set a value to `accepted_*` so, it will be clear then that `0` actually means nothing was accepted. 

If the server doesn't follow this, and initializes the `partial_success` anyway, at least we will have `accepted_` as not set/null, as marking it as `optional` enables that in proto3.

With this I think we don't need to touch the `error_message`. We discussed here a lot that this message is probably not that useful to receivers so I feel we shouldn't use it's state to any logic definition of what actually happened. 

I also feel we might be talking pass each other and it's tricky to convey all this with words 😅. I'm up for a live discussion, be it in the next spec call or via Slack. I feel that will be more productive. What do you think?

Reply:
I went with @jmacd option (2)

> Make each accepted_* field optional

This way, if the server doesn't set it, it will be possible to distinguish the 'not set' value from the default one, `0`. I also extended the documentation on the field to make this clear. 

I will mark this conversation as resolved.